Add Repo Contents API endpoints.

scott-fleischman opened the pull request:
This adds the create, update and delete file endpoints from https://developer.github.com/v3/repos/contents/

The code in the PR works great.

I would say it's about 80% done in terms of the coding guidelines. Still to be done are updating the contents example from the samples (which is also out of date with respect to the current master branch). I wanted to wait to update those until I got approval on the current code structure.

I am opening the Pull Request now to get feedback on some of the code organizational choices and naming choices. I am happy to change them to whatever you think best. I am not particularly attached to my choices so far--I mainly picked what seemed to be a reasonably good start.

In particular I added a new module called `GitHub.Endpoints.Repos.Contents`. This mirrors the structure of the GitHub API.

I also moved `contentFor` and related functions to the new module, since they are located there in the GitHub API.

I added the new endpoints to the same module.

I added new datatypes to the module `GitHub.Data.Content` because that seemed like the best fit so far.

scott-fleischman commented:
One further design consideration: the content fields for `CreateFile` and `UpdateFIle` are required to be in base64 encoding.

Right now there is nothing in this API that clues you in to that fact--you have to read the API documentation.

Should we do one of the following?
1. Change the name of the field to something like `contentBase64`
2. Automatically do the base64 encoding (perhaps using the `base64-bytestring` package).

scott-fleischman commented:
@phadej do you have any initial feedback on this PR?

phadej commented on src/GitHub/Data/Content.hs:
```diff
@@ -5,6 +5,9 @@
 --
 module GitHub.Data.Content where
 
+import Data.Aeson.Types        (KeyValue)
```
we'll need to bump lower bump on aeson for this

phadej commented on src/GitHub/Data/Content.hs:
```diff
@@ -87,3 +155,74 @@ instance FromJSON ContentInfo where
+                 ]
+
+instance ToJSON CreateFile where
+  toJSON (CreateFile { createFilePath       = path
```
I'd use `RecordWildCards`, this is extremely ugly. (No need to touch old code though)

phadej commented on src/GitHub/Data/Content.hs:
```diff
@@ -87,3 +155,74 @@ instance FromJSON ContentInfo where
+                     ++ "author"           .=? author
+                     ++ "committer"        .=? committer
+
+(.=?) :: ToJSON v => KeyValue t => Text -> Maybe v -> [t]
```
cool!

phadej replied:
With exolicit `[Aeson.Pair]` this will be compatible with old aesons, though we can bump lower bump too.

scott-fleischman commented:
@phadej I updated the code based on your comments.

I also updated the Contents sample code so that it builds. I added it to the cabal file so it builds as part of github-samples. Finally, I added an example using the new create/update/delete file code.

Let me know if you would like anything else done on this PR.

phadej commented:
Great, thanks!
